File: shared/src/commonTest/kotlin/com/ssbmax/shared/domain/model/FeatureFlagsTest.kt
package com.ssbmax.shared.domain.model

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * [isAppVersionBelowMinimum] is Phase 8's actual gate decision -- every other
 * moving part (repository, ViewModel, screen) exists to feed it two strings
 * and act on the result, so its comparison semantics are worth pinning
 * directly rather than only indirectly through [AppRootViewModel]'s tests.
 */
class FeatureFlagsTest {

    @Test
    fun `equal versions are not below minimum`() {
        assertFalse(isAppVersionBelowMinimum(current = "1.2.3", minimum = "1.2.3"))
    }

    @Test
    fun `a higher current version is not below minimum`() {
        assertFalse(isAppVersionBelowMinimum(current = "2.0.0", minimum = "1.9.9"))
    }

    @Test
    fun `a lower current version is below minimum`() {
        assertTrue(isAppVersionBelowMinimum(current = "1.0.0", minimum = "1.0.1"))
    }

    @Test
    fun `compares numerically rather than lexicographically`() {
        // Lexicographic compare would say "9" > "10" -- a real regression class for version strings.
        assertTrue(isAppVersionBelowMinimum(current = "1.9.0", minimum = "1.10.0"))
    }

    @Test
    fun `a shorter version string is padded with zero segments`() {
        assertFalse(isAppVersionBelowMinimum(current = "1.2", minimum = "1.2.0"))
        assertTrue(isAppVersionBelowMinimum(current = "1.2", minimum = "1.2.1"))
    }

    @Test
    fun `a malformed segment is treated as zero instead of throwing`() {
        assertEquals(false, isAppVersionBelowMinimum(current = "1.x.0", minimum = "1.0.0"))
    }

    @Test
    fun `SAFE_DEFAULT carries the compiled-in contract minimum and no flags`() {
        assertEquals(
            com.ssbmax.shared.contracts.SsbContracts.Routes.MINIMUM_SUPPORTED_APP_VERSION,
            FeatureFlags.SAFE_DEFAULT.minimumSupportedAppVersion
        )
        assertTrue(FeatureFlags.SAFE_DEFAULT.flags.isEmpty())
    }

    @Test
    fun `isEnabled falls back to the caller-supplied default for an unknown flag`() {
        val flags = FeatureFlags(flags = mapOf("known" to true))

        assertTrue(flags.isEnabled("known"))
        assertFalse(flags.isEnabled("unknown"))
        assertTrue(flags.isEnabled("unknown", default = true))
    }
}

File: web/tests/unit/featureFlags.test.ts
import { describe, it, expect } from 'vitest';
import { isAppVersionBelowMinimum, SAFE_DEFAULT_FEATURE_FLAGS } from '../../src/types/featureFlags';
import { Routes } from '../../src/generated/contracts';

/**
 * `isAppVersionBelowMinimum` is Phase 8's actual gate decision (mirrors
 * KMP's `FeatureFlagsTest.kt`) -- pinned directly since every other moving
 * part (repository, hook, screen) exists only to feed it two strings.
 */
describe('isAppVersionBelowMinimum', () => {
  it('treats equal versions as not below minimum', () => {
    expect(isAppVersionBelowMinimum('1.2.3', '1.2.3')).toBe(false);
  });

  it('treats a higher current version as not below minimum', () => {
    expect(isAppVersionBelowMinimum('2.0.0', '1.9.9')).toBe(false);
  });

  it('treats a lower current version as below minimum', () => {
    expect(isAppVersionBelowMinimum('1.0.0', '1.0.1')).toBe(true);
  });

  it('compares numerically, not lexicographically', () => {
    // Lexicographic compare would say "9" > "10" -- a real regression class for version strings.
    expect(isAppVersionBelowMinimum('1.9.0', '1.10.0')).toBe(true);
  });

  it('pads a shorter version string with zero segments', () => {
    expect(isAppVersionBelowMinimum('1.2', '1.2.0')).toBe(false);
    expect(isAppVersionBelowMinimum('1.2', '1.2.1')).toBe(true);
  });

  it('treats a malformed segment as zero instead of throwing', () => {
    expect(isAppVersionBelowMinimum('1.x.0', '1.0.0')).toBe(false);
  });
});

describe('SAFE_DEFAULT_FEATURE_FLAGS', () => {
  it('carries the compiled-in contract minimum and no flags', () => {
    expect(SAFE_DEFAULT_FEATURE_FLAGS.minimumSupportedAppVersion).toBe(Routes.MINIMUM_SUPPORTED_APP_VERSION);
    expect(SAFE_DEFAULT_FEATURE_FLAGS.flags).toEqual({});
  });
});
